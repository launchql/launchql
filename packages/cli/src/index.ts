#!/usr/bin/env node
import { CLI, CLIOptions } from 'inquirerer';

import { commands } from './commands';

export const options: Partial<CLIOptions> = {
  minimistOpts: {
    alias: {
      v: 'version',
      h: 'help'
    }
  }
};

const app = new CLI(commands, options);

app.run().then(()=> {
  // all done!
}).catch(error => {
  // Should not reach here with the new CLI error handling pattern
  // But keep as fallback for unexpected errors
  console.error('Unexpected error:', error);
  process.exit(1);
});