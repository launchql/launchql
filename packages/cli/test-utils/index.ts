export * from './cli';
export * from './fixtures';
export * from './CLIDeployTestFixture';
export * from './TestDatabase';
