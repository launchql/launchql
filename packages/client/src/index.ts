import { Pool, PoolClient } from 'pg';

import env from './env';

export class Database {
  private static instance: Database;
  private pool: Pool;

  constructor() {
    if (Database.instance) {
      return Database.instance;
    }

    const pgPoolConfig = {
      connectionString: env.DATABASE_URL,
    };
    this.pool = new Pool(pgPoolConfig);

    // Ensure the pool is closed on process termination
    process.on('SIGTERM', async () => {
      await this.shutdown();
    });

    Database.instance = this;
    return this;
  }

  /**
   * Executes a callback function within a database transaction.
   * @param fn - A callback function that receives a PoolClient to perform database operations.
   */
  async withTransaction(fn: (client: PoolClient) => Promise<void>): Promise<void> {
    const client = await this.pool.connect();
    try {
      await client.query('BEGIN');
      try {
        await fn(client);
        await client.query('COMMIT');
      } catch (e) {
        console.error('Error during transaction:', e);
        await client.query('ROLLBACK');
        throw e; // Re-throw the error to propagate it
      }
    } finally {
      client.release();
    }
  }

  /**
   * Shuts down the connection pool.
   */
  async shutdown(): Promise<void> {
    await this.pool.end();
  }
}