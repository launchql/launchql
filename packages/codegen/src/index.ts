import { generateCode } from './codegen';
import getIntrospectionRows, { GetIntrospectionRowsOptions } from './introspect';
import { DatabaseObject } from './types';

(async () => {
  const options: GetIntrospectionRowsOptions = {
    connectionString: 'postgresql://postgres:password@localhost:5432/mydb',
    introspectionOptions: {
      pgLegacyFunctionsOnly: false,
      pgIgnoreRBAC: true,
    },
    namespacesToIntrospect: ['collections_public'],
    includeExtensions: false,
  };

  try {
    // Fetch introspection rows
    const rows: DatabaseObject[] = await getIntrospectionRows(options);
    console.log('Introspection Rows Fetched:', rows);

    // Generate TypeScript code
    const codegenOptions = {
      includeTimestamps: true,
      includeUUID: true,
    };
    const generatedCode = generateCode(rows, codegenOptions);

    console.log('Generated TypeScript Code:');
    console.log(generatedCode);
  } catch (error) {
    console.error('Failed to fetch introspection rows or generate code:', error);
  }
})();
