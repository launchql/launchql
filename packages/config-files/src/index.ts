export * from './types';
export * from './plan';
export * from './config';
export * from './sql';
export * from './extension';
export * from './sql-scripts';