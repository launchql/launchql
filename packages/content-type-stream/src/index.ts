export * from './content-stream';
export * from './content-type-stream';