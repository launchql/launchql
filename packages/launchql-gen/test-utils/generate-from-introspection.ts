import { IntrospectionQueryResult, parseGraphQuery } from 'introspectron';
import { print } from 'graphql';

import { generate, GqlMap } from '../src/gql';

export function generateFromIntrospection(introspection: IntrospectionQueryResult): Record<string, string> {
    const { queries, mutations } = parseGraphQuery(introspection);
    const gqlMap: GqlMap = { ...queries, ...mutations };
    const gen = generate(gqlMap);
  
    return Object.entries(gen).reduce<Record<string, string>>((acc, [key, val]) => {
      if (val?.ast) {
        acc[key] = print(val.ast);
      }
      return acc;
    }, {});
  }