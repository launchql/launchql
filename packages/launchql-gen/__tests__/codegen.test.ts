process.env.LOG_SCOPE = 'launchql-codegen';

import { getConnections, GraphQLQueryFn, seed } from 'graphile-test';
import { IntrospectionQuery, IntrospectionQueryResult, parseGraphQuery } from 'introspectron';
import { print } from 'graphql';

import { join } from 'path';
import { generate, GqlMap } from '../src';
import { generateFromIntrospection } from '../test-utils/generate-from-introspection';

const sql = (f: string) => join(__dirname, '/../sql', f);

let teardown: () => Promise<void>;
let query: GraphQLQueryFn;
let introspection: IntrospectionQueryResult;

beforeAll(async () => {
  ({ query, teardown } = await getConnections(
    {
      schemas: ['launchql_gen']
    },
    [seed.sqlfile([sql('test.sql')])]
  ));
});

afterAll(() => teardown());

beforeAll(async () => {
  const res = await query(IntrospectionQuery);

  if (res.errors) {
    console.error(res.errors);
    throw new Error('GraphQL introspection query failed');
  }

  introspection = res.data as IntrospectionQueryResult;
});

it('generates output', () => {

  const { queries, mutations } = parseGraphQuery(introspection);
  const gqlMap: GqlMap = { ...queries, ...mutations };
  const gen = generate(gqlMap);
  const output = Object.keys(gen).reduce<Record<string, string>>((acc, key) => {
    const entry = gen[key];
    if (entry?.ast) {
      // @ts-ignore
      acc[key] = print(entry.ast);
    }
    return acc;
  }, {});

  expect(output).toMatchSnapshot();

});

it('helper method', () => {
  expect(introspection).toBeDefined();
  const output = generateFromIntrospection(introspection);
  expect(output).toMatchSnapshot();
});