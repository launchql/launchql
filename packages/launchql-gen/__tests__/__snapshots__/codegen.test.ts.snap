// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates output 1`] = `
{
  "activeUserFragment": "fragment activeUserFragment on ActiveUser {
  id
  username
}
",
  "createActiveUserMutation": "mutation createActiveUserMutation($username: String) {
  createActiveUser(input: {activeUser: {username: $username}}) {
    activeUser {
      id
      username
    }
  }
}
",
  "createPostMutation": "mutation createPostMutation($userId: Int!, $title: String!, $body: String, $published: Boolean, $publishedAt: Datetime) {
  createPost(
    input: {post: {userId: $userId, title: $title, body: $body, published: $published, publishedAt: $publishedAt}}
  ) {
    post {
      id
      userId
      title
      body
      published
      publishedAt
    }
  }
}
",
  "createUserMutation": "mutation createUserMutation($username: String!, $email: String) {
  createUser(input: {user: {username: $username, email: $email}}) {
    user {
      id
      username
      email
      createdAt
    }
  }
}
",
  "deletePostMutation": "mutation deletePostMutation($id: UUID!) {
  deletePost(input: {id: $id}) {
    clientMutationId
  }
}
",
  "deleteUserByUsernameMutation": "mutation deleteUserByUsernameMutation($username: String!) {
  deleteUserByUsername(input: {username: $username}) {
    clientMutationId
  }
}
",
  "deleteUserMutation": "mutation deleteUserMutation($id: Int!) {
  deleteUser(input: {id: $id}) {
    clientMutationId
  }
}
",
  "getActiveUsersOrderByEnums": "query getActiveUsersOrderByEnums {
  __type(name: "ActiveUsersOrderBy") {
    enumValues {
      name
    }
  }
}
",
  "getActiveUsersPaginated": "query getActiveUsersPaginated($first: Int, $last: Int, $offset: Int, $after: Cursor, $before: Cursor, $condition: ActiveUserCondition, $filter: ActiveUserFilter, $orderBy: [ActiveUsersOrderBy!]) {
  activeUsers(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    edges {
      cursor
      node {
        id
        username
      }
    }
  }
}
",
  "getActiveUsersQuery": "query getActiveUsersQuery($first: Int, $last: Int, $after: Cursor, $before: Cursor, $offset: Int, $condition: ActiveUserCondition, $filter: ActiveUserFilter, $orderBy: [ActiveUsersOrderBy!]) {
  activeUsers(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    nodes {
      id
      username
    }
  }
}
",
  "getActiveUsersQueryAll": "query getActiveUsersQueryAll {
  activeUsers {
    totalCount
    nodes {
      id
      username
    }
  }
}
",
  "getMetaQuery": "query getMetaQuery {
  _meta {
    tables(first: 3) {
      nodes {
        name
        query
        inflection
        relations
        fields
        constraints
        foreignKeyConstraints
        primaryKeyConstraints
        uniqueConstraints
        checkConstraints
        exclusionConstraints
      }
    }
  }
}
",
  "getPostQuery": "query getPostQuery($id: UUID!) {
  post(id: $id) {
    id
    userId
    title
    body
    published
    publishedAt
    user
  }
}
",
  "getPostsOrderByEnums": "query getPostsOrderByEnums {
  __type(name: "PostsOrderBy") {
    enumValues {
      name
    }
  }
}
",
  "getPostsPaginated": "query getPostsPaginated($first: Int, $last: Int, $offset: Int, $after: Cursor, $before: Cursor, $condition: PostCondition, $filter: PostFilter, $orderBy: [PostsOrderBy!]) {
  posts(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    edges {
      cursor
      node {
        id
        userId
        title
        body
        published
        publishedAt
        user
      }
    }
  }
}
",
  "getPostsQuery": "query getPostsQuery($first: Int, $last: Int, $after: Cursor, $before: Cursor, $offset: Int, $condition: PostCondition, $filter: PostFilter, $orderBy: [PostsOrderBy!]) {
  posts(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    nodes {
      id
      userId
      title
      body
      published
      publishedAt
      user
    }
  }
}
",
  "getPostsQueryAll": "query getPostsQueryAll {
  posts {
    totalCount
    nodes {
      id
      userId
      title
      body
      published
      publishedAt
      user
    }
  }
}
",
  "getUserByUsernameQuery": "query getUserByUsernameQuery($username: String!) {
  userByUsername(username: $username) {
    id
    username
    email
    createdAt
    posts(first: 3) {
      nodes {
        id
        userId
        title
        body
        published
        publishedAt
        user
      }
    }
  }
}
",
  "getUserQuery": "query getUserQuery($id: Int!) {
  user(id: $id) {
    id
    username
    email
    createdAt
    posts(first: 3) {
      nodes {
        id
        userId
        title
        body
        published
        publishedAt
        user
      }
    }
  }
}
",
  "getUsersOrderByEnums": "query getUsersOrderByEnums {
  __type(name: "UsersOrderBy") {
    enumValues {
      name
    }
  }
}
",
  "getUsersPaginated": "query getUsersPaginated($first: Int, $last: Int, $offset: Int, $after: Cursor, $before: Cursor, $condition: UserCondition, $filter: UserFilter, $orderBy: [UsersOrderBy!]) {
  users(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    edges {
      cursor
      node {
        id
        username
        email
        createdAt
        posts(first: 3) {
          nodes {
            id
            userId
            title
            body
            published
            publishedAt
            user
          }
        }
      }
    }
  }
}
",
  "getUsersQuery": "query getUsersQuery($first: Int, $last: Int, $after: Cursor, $before: Cursor, $offset: Int, $condition: UserCondition, $filter: UserFilter, $orderBy: [UsersOrderBy!]) {
  users(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    nodes {
      id
      username
      email
      createdAt
      posts(first: 3) {
        nodes {
          id
          userId
          title
          body
          published
          publishedAt
          user
        }
      }
    }
  }
}
",
  "getUsersQueryAll": "query getUsersQueryAll {
  users {
    totalCount
    nodes {
      id
      username
      email
      createdAt
      posts(first: 3) {
        nodes {
          id
          userId
          title
          body
          published
          publishedAt
          user
        }
      }
    }
  }
}
",
  "postFragment": "fragment postFragment on Post {
  id
  userId
  title
  body
  published
  publishedAt
  user
}
",
  "updatePostMutation": "mutation updatePostMutation($userId: Int, $title: String, $body: String, $published: Boolean, $publishedAt: Datetime) {
  updatePost(
    input: {id: $id, patch: {userId: $userId, title: $title, body: $body, published: $published, publishedAt: $publishedAt}}
  ) {
    post {
      id
      userId
      title
      body
      published
      publishedAt
    }
  }
}
",
  "updateUserByUsernameMutation": "mutation updateUserByUsernameMutation($username: String!, $email: String) {
  updateUserByUsername(input: {username: $username, patch: {email: $email}}) {
    user {
      id
      username
      email
      createdAt
    }
  }
}
",
  "updateUserMutation": "mutation updateUserMutation($username: String, $email: String) {
  updateUser(input: {id: $id, patch: {username: $username, email: $email}}) {
    user {
      id
      username
      email
      createdAt
    }
  }
}
",
  "userFragment": "fragment userFragment on User {
  id
  username
  email
  createdAt
  [object Object]
}
",
}
`;

exports[`helper method 1`] = `
{
  "activeUserFragment": "fragment activeUserFragment on ActiveUser {
  id
  username
}
",
  "createActiveUserMutation": "mutation createActiveUserMutation($username: String) {
  createActiveUser(input: {activeUser: {username: $username}}) {
    activeUser {
      id
      username
    }
  }
}
",
  "createPostMutation": "mutation createPostMutation($userId: Int!, $title: String!, $body: String, $published: Boolean, $publishedAt: Datetime) {
  createPost(
    input: {post: {userId: $userId, title: $title, body: $body, published: $published, publishedAt: $publishedAt}}
  ) {
    post {
      id
      userId
      title
      body
      published
      publishedAt
    }
  }
}
",
  "createUserMutation": "mutation createUserMutation($username: String!, $email: String) {
  createUser(input: {user: {username: $username, email: $email}}) {
    user {
      id
      username
      email
      createdAt
    }
  }
}
",
  "deletePostMutation": "mutation deletePostMutation($id: UUID!) {
  deletePost(input: {id: $id}) {
    clientMutationId
  }
}
",
  "deleteUserByUsernameMutation": "mutation deleteUserByUsernameMutation($username: String!) {
  deleteUserByUsername(input: {username: $username}) {
    clientMutationId
  }
}
",
  "deleteUserMutation": "mutation deleteUserMutation($id: Int!) {
  deleteUser(input: {id: $id}) {
    clientMutationId
  }
}
",
  "getActiveUsersOrderByEnums": "query getActiveUsersOrderByEnums {
  __type(name: "ActiveUsersOrderBy") {
    enumValues {
      name
    }
  }
}
",
  "getActiveUsersPaginated": "query getActiveUsersPaginated($first: Int, $last: Int, $offset: Int, $after: Cursor, $before: Cursor, $condition: ActiveUserCondition, $filter: ActiveUserFilter, $orderBy: [ActiveUsersOrderBy!]) {
  activeUsers(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    edges {
      cursor
      node {
        id
        username
      }
    }
  }
}
",
  "getActiveUsersQuery": "query getActiveUsersQuery($first: Int, $last: Int, $after: Cursor, $before: Cursor, $offset: Int, $condition: ActiveUserCondition, $filter: ActiveUserFilter, $orderBy: [ActiveUsersOrderBy!]) {
  activeUsers(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    nodes {
      id
      username
    }
  }
}
",
  "getActiveUsersQueryAll": "query getActiveUsersQueryAll {
  activeUsers {
    totalCount
    nodes {
      id
      username
    }
  }
}
",
  "getMetaQuery": "query getMetaQuery {
  _meta {
    tables(first: 3) {
      nodes {
        name
        query
        inflection
        relations
        fields
        constraints
        foreignKeyConstraints
        primaryKeyConstraints
        uniqueConstraints
        checkConstraints
        exclusionConstraints
      }
    }
  }
}
",
  "getPostQuery": "query getPostQuery($id: UUID!) {
  post(id: $id) {
    id
    userId
    title
    body
    published
    publishedAt
    user
  }
}
",
  "getPostsOrderByEnums": "query getPostsOrderByEnums {
  __type(name: "PostsOrderBy") {
    enumValues {
      name
    }
  }
}
",
  "getPostsPaginated": "query getPostsPaginated($first: Int, $last: Int, $offset: Int, $after: Cursor, $before: Cursor, $condition: PostCondition, $filter: PostFilter, $orderBy: [PostsOrderBy!]) {
  posts(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    edges {
      cursor
      node {
        id
        userId
        title
        body
        published
        publishedAt
        user
      }
    }
  }
}
",
  "getPostsQuery": "query getPostsQuery($first: Int, $last: Int, $after: Cursor, $before: Cursor, $offset: Int, $condition: PostCondition, $filter: PostFilter, $orderBy: [PostsOrderBy!]) {
  posts(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    nodes {
      id
      userId
      title
      body
      published
      publishedAt
      user
    }
  }
}
",
  "getPostsQueryAll": "query getPostsQueryAll {
  posts {
    totalCount
    nodes {
      id
      userId
      title
      body
      published
      publishedAt
      user
    }
  }
}
",
  "getUserByUsernameQuery": "query getUserByUsernameQuery($username: String!) {
  userByUsername(username: $username) {
    id
    username
    email
    createdAt
    posts(first: 3) {
      nodes {
        id
        userId
        title
        body
        published
        publishedAt
        user
      }
    }
  }
}
",
  "getUserQuery": "query getUserQuery($id: Int!) {
  user(id: $id) {
    id
    username
    email
    createdAt
    posts(first: 3) {
      nodes {
        id
        userId
        title
        body
        published
        publishedAt
        user
      }
    }
  }
}
",
  "getUsersOrderByEnums": "query getUsersOrderByEnums {
  __type(name: "UsersOrderBy") {
    enumValues {
      name
    }
  }
}
",
  "getUsersPaginated": "query getUsersPaginated($first: Int, $last: Int, $offset: Int, $after: Cursor, $before: Cursor, $condition: UserCondition, $filter: UserFilter, $orderBy: [UsersOrderBy!]) {
  users(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    edges {
      cursor
      node {
        id
        username
        email
        createdAt
        posts(first: 3) {
          nodes {
            id
            userId
            title
            body
            published
            publishedAt
            user
          }
        }
      }
    }
  }
}
",
  "getUsersQuery": "query getUsersQuery($first: Int, $last: Int, $after: Cursor, $before: Cursor, $offset: Int, $condition: UserCondition, $filter: UserFilter, $orderBy: [UsersOrderBy!]) {
  users(
    first: $first
    last: $last
    offset: $offset
    after: $after
    before: $before
    condition: $condition
    filter: $filter
    orderBy: $orderBy
  ) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
      startCursor
    }
    nodes {
      id
      username
      email
      createdAt
      posts(first: 3) {
        nodes {
          id
          userId
          title
          body
          published
          publishedAt
          user
        }
      }
    }
  }
}
",
  "getUsersQueryAll": "query getUsersQueryAll {
  users {
    totalCount
    nodes {
      id
      username
      email
      createdAt
      posts(first: 3) {
        nodes {
          id
          userId
          title
          body
          published
          publishedAt
          user
        }
      }
    }
  }
}
",
  "postFragment": "fragment postFragment on Post {
  id
  userId
  title
  body
  published
  publishedAt
  user
}
",
  "updatePostMutation": "mutation updatePostMutation($userId: Int, $title: String, $body: String, $published: Boolean, $publishedAt: Datetime) {
  updatePost(
    input: {id: $id, patch: {userId: $userId, title: $title, body: $body, published: $published, publishedAt: $publishedAt}}
  ) {
    post {
      id
      userId
      title
      body
      published
      publishedAt
    }
  }
}
",
  "updateUserByUsernameMutation": "mutation updateUserByUsernameMutation($username: String!, $email: String) {
  updateUserByUsername(input: {username: $username, patch: {email: $email}}) {
    user {
      id
      username
      email
      createdAt
    }
  }
}
",
  "updateUserMutation": "mutation updateUserMutation($username: String, $email: String) {
  updateUser(input: {id: $id, patch: {username: $username, email: $email}}) {
    user {
      id
      username
      email
      createdAt
    }
  }
}
",
  "userFragment": "fragment userFragment on User {
  id
  username
  email
  createdAt
  [object Object]
}
",
}
`;
